File: Easy/JAVASCRIPT/ClosestValueInBST.js

function getClosetValueInBST(node, target) {
    if (node === null) return;
    return getClosetValueInBSTHelper(node, target, node.value);
}

function getClosetValueInBSTHelper(node, target, closest) {
    if (node === null) return closest;
    if (Math.abs(target - node.value) < Math.abs(target - closest)) {
        closest = node.value;
    }
    if (target < node.value) {
        return getClosetValueInBSTHelper(node.left, target, closest);
    }
    else if (target > node.value) {
        return getClosetValueInBSTHelper(node.right, target, closest);
    }
    else {
        return closest;
    }
}

class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

const root = new Node(10);
root.left = new Node(5);
root.left.left = new Node(2);
root.left.left.left = new Node(1);
root.left.right = new Node(5);
root.right = new Node(15);
root.right.left = new Node(13);
root.right.left.right = new Node(14);
root.right.right = new Node(22);


let result = getClosetValueInBST(root, 11);
console.log(result);

File: Easy/JAVA/ClosestValueInBST.java

class ClosestValueInBST {

    public class Node {
        public int value;
        public Node left;
        public Node right;

        public Node(int value) {
            this.value = value;
        }

    }

    public static Node root;

    public void insert(int value)
    {
        var node = new Node(value);
        if (root == null)
        {
            root = node;
            return;
        }

        var current = root;
        while (true)
        {
            if (value < current.value)
            {
                if (current.left == null)
                {
                    current.left = node;
                    break;
                }
                current = current.left;
            }
            else
            {
                if (current.right == null)
                {
                    current.right = node;
                    break;
                }
                current = current.right;
            }
        }
    }

    public static int findClosestValueInBST(Node node, int target) {
        return findClosestValueInBST(node, target, root.value);
    }

    public static int findClosestValueInBST(Node node, int target, int closest) {
        if (Math.abs(target - node.value) < Math.abs(target - closest))
            closest = node.value;
        if (target < node.value && node.left != null)
            return findClosestValueInBST(node.left, target, closest);
        else if (target > node.value && node.right != null)
            return findClosestValueInBST(node.right, target, closest);
        else
            return closest;
    }

    public static void main(String[] args) {
        var tree = new ClosestValueInBST();
        var values = new int[] { 10, 5, 2, 1, 5, 15, 13, 14, 22 };
        for (var value : values)
        {
        tree.insert(value);
        }

        var result = findClosestValueInBST(ClosestValueInBST.root, 20);
        System.out.println(result);
       
    }
}